### PR TITLE
Consolidate error messages

### PR DESCRIPTION
**Pull request by tompsota**
Makes the error output prettier by removing
unnecessary parts, e.g. traceback, full yaml and
fixes special character escaping issue in errors.

Signed-off-by: Tomas Psota <tpsota@redhat.com>

**Is this a BUG FIX or a FEATURE ?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
> /kind enhancement

**What this PR does / why we need it**:

Makes the error output of NNCE error message
prettier by removing unnecessary parts, e.g. traceback,
 full yaml, and ixes special character escaping issue in errors.

**Output samples:**

_Example 1_

**Original**
```
message: |-
      error reconciling NodeNetworkConfigurationPolicy at desired state apply: ,
       failed to execute nmstatectl set --no-commit --timeout 480: 'exit status 1' '' '2021-04-15 08:17:35,030 root         WARNING  libnm version 1.26.7 mismatches NetworkManager version 1.29.9
      2021-04-15 08:17:35,057 root         DEBUG    Async action: Create checkpoint started
      2021-04-15 08:17:35,060 root         DEBUG    Checkpoint None created for all devices
      2021-04-15 08:17:35,060 root         DEBUG    Async action: Create checkpoint finished
      2021-04-15 08:17:35,061 root         DEBUG    Async action: Add profile: eth666 started
      2021-04-15 08:17:35,065 root         DEBUG    Async action: Add profile: eth666 finished
      2021-04-15 08:17:35,065 root         DEBUG    Async action: Activate profile: eth666 started
      2021-04-15 08:17:35,066 root         DEBUG    Action Activate profile: eth666 failed, trying again.
      2021-04-15 08:17:35,067 root         DEBUG    Async action: Rollback to checkpoint /org/freedesktop/NetworkManager/Checkpoint/2 started
      2021-04-15 08:17:35,069 root         DEBUG    Checkpoint /org/freedesktop/NetworkManager/Checkpoint/2 rollback executed
      2021-04-15 08:17:35,069 root         DEBUG    Interface lo rollback succeeded
      2021-04-15 08:17:35,069 root         DEBUG    Interface calic82870b62e9 rollback succeeded
      2021-04-15 08:17:35,069 root         DEBUG    Interface eth0 rollback succeeded
      2021-04-15 08:17:35,069 root         DEBUG    Interface cali661fbd931aa rollback succeeded
      2021-04-15 08:17:35,069 root         DEBUG    Interface cali7d7f8e3e762 rollback succeeded
      2021-04-15 08:17:35,069 root         DEBUG    Interface eth2 rollback succeeded
      2021-04-15 08:17:35,069 root         DEBUG    Interface cali770b5dc9800 rollback succeeded
      2021-04-15 08:17:35,069 root         DEBUG    Interface docker0 rollback succeeded
      2021-04-15 08:17:35,069 root         DEBUG    Interface cali3a2807641ac rollback succeeded
      2021-04-15 08:17:35,069 root         DEBUG    Interface tunl0 rollback succeeded
      2021-04-15 08:17:35,069 root         DEBUG    Interface cali8ffbc7dfd36 rollback succeeded
      2021-04-15 08:17:35,070 root         DEBUG    Interface eth1 rollback succeeded
      2021-04-15 08:17:35,070 root         DEBUG    Interface cali9e8a4e2983a rollback succeeded
      2021-04-15 08:17:35,070 root         DEBUG    Interface cali074cadeaf8a rollback succeeded
      2021-04-15 08:17:35,070 root         DEBUG    Async action: Rollback to checkpoint /org/freedesktop/NetworkManager/Checkpoint/2 finished
      Traceback (most recent call last):
        File "/usr/bin/nmstatectl", line 11, in <module>
          load_entry_point('nmstate==0.3.7', 'console_scripts', 'nmstatectl')()
        File "/usr/lib/python3.6/site-packages/nmstatectl/nmstatectl.py", line 69, in main
          return args.func(args)
        File "/usr/lib/python3.6/site-packages/nmstatectl/nmstatectl.py", line 279, in apply
          args.save_to_disk,
        File "/usr/lib/python3.6/site-packages/nmstatectl/nmstatectl.py", line 308, in apply_state
          save_to_disk=save_to_disk,
        File "/usr/lib/python3.6/site-packages/libnmstate/netapplier.py", line 73, in apply
          _apply_ifaces_state(plugins, net_state, verify_change, save_to_disk)
        File "/usr/lib/python3.6/site-packages/libnmstate/netapplier.py", line 106, in _apply_ifaces_state
          plugin.apply_changes(net_state, save_to_disk)
        File "/usr/lib/python3.6/site-packages/libnmstate/nm/plugin.py", line 174, in apply_changes
          nm_applier.apply_changes(self.context, net_state, save_to_disk)
        File "/usr/lib/python3.6/site-packages/libnmstate/nm/applier.py", line 183, in apply_changes
          _set_ifaces_admin_state(context, ifaces_desired_state, con_profiles)
        File "/usr/lib/python3.6/site-packages/libnmstate/nm/applier.py", line 354, in _set_ifaces_admin_state
          context.wait_all_finish()
        File "/usr/lib/python3.6/site-packages/libnmstate/nm/context.py", line 215, in wait_all_finish
          raise tmp_error
      libnmstate.error.NmstateLibnmError: Activate profile: eth666 failed: error=nm-manager-error-quark: No suitable device found for this connection (device eth0 not available because profile is not compatible with device (mismatching interface name)). (3)
      '
```

**Current (python traceback, debug and warning removed)**
```
    message: |
      error reconciling NodeNetworkConfigurationPolicy at desired state apply: ,
       failed to execute nmstatectl set --no-commit --timeout 480: 'exit status 1'
      libnmstate.error.NmstateLibnmError
        Activate profile: eth666 failed
          error=nm-manager-error-quark
            No suitable device found for this connection (device eth0 not available because profile is not compatible with device (mismatching interface name)). (3)
```

_Example 2_

**Original**
```
error reconciling NodeNetworkConfigurationPolicy at desired state apply:
      ,\n failed to execute nmstatectl set --no-commit --timeout 480: 'exit status
      1' '' '2021-04-15 08:14:08,025 root         WARNING  libnm version 1.26.7 mismatches
      NetworkManager version 1.29.9\n2021-04-15 08:14:08,052 root         DEBUG    Async
      action: Create checkpoint started\n2021-04-15 08:14:08,059 root         DEBUG
      \   Checkpoint None created for all devices\n2021-04-15 08:14:08,059 root         DEBUG
      \   Async action: Create checkpoint finished\n2021-04-15 08:14:08,060 root         DEBUG
      \   Async action: Add profile: eth1 started\n2021-04-15 08:14:08,064 root         DEBUG
      \   Async action: Add profile: eth1 finished\n2021-04-15 08:14:08,065 root         DEBUG
      \   Async action: Activate profile: eth1 started\n2021-04-15 08:14:08,067 root
      \        DEBUG    Connection activation initiated: dev=eth1, con-state=<enum
      NM_ACTIVE_CONNECTION_STATE_ACTIVATING of type NM.ActiveConnectionState>\n2021-04-15
      08:14:08,082 root         DEBUG    Connection activation succeeded: dev=eth1,
      con-state=<enum NM_ACTIVE_CONNECTION_STATE_ACTIVATED of type NM.ActiveConnectionState>,
      dev-state=<enum NM_DEVICE_STATE_ACTIVATED of type NM.DeviceState>, state-flags=<flags
      NM_ACTIVATION_STATE_FLAG_LAYER2_READY | NM_ACTIVATION_STATE_FLAG_IP4_READY |
      NM_ACTIVATION_STATE_FLAG_IP6_READY of type NM.ActivationStateFlags>\n2021-04-15
      08:14:08,082 root         DEBUG    Async action: Activate profile: eth1 finished\n2021-04-15
      08:14:28,774 root         DEBUG    Async action: Rollback to checkpoint /org/freedesktop/NetworkManager/Checkpoint/1
      started\n2021-04-15 08:14:28,779 root         DEBUG    Checkpoint /org/freedesktop/NetworkManager/Checkpoint/1
      rollback executed\n2021-04-15 08:14:28,780 root         DEBUG    Interface cali661fbd931aa
      rollback succeeded\n2021-04-15 08:14:28,780 root         DEBUG    Interface
      cali7d7f8e3e762 rollback succeeded\n2021-04-15 08:14:28,780 root         DEBUG
      \   Interface cali3a2807641ac rollback succeeded\n2021-04-15 08:14:28,780 root
      \        DEBUG    Interface cali074cadeaf8a rollback succeeded\n2021-04-15 08:14:28,780
      root         DEBUG    Interface cali9e8a4e2983a rollback succeeded\n2021-04-15
      08:14:28,780 root         DEBUG    Interface docker0 rollback succeeded\n2021-04-15
      08:14:28,780 root         DEBUG    Interface eth2 rollback succeeded\n2021-04-15
      08:14:28,780 root         DEBUG    Interface cali8ffbc7dfd36 rollback succeeded\n2021-04-15
      08:14:28,780 root         DEBUG    Interface cali770b5dc9800 rollback succeeded\n2021-04-15
      08:14:28,780 root         DEBUG    Interface tunl0 rollback succeeded\n2021-04-15
      08:14:28,780 root         DEBUG    Interface eth1 rollback succeeded\n2021-04-15
      08:14:28,780 root         DEBUG    Interface eth0 rollback succeeded\n2021-04-15
      08:14:28,780 root         DEBUG    Interface lo rollback succeeded\n2021-04-15
      08:14:28,780 root         DEBUG    Interface calic82870b62e9 rollback succeeded\n2021-04-15
      08:14:28,780 root         DEBUG    Async action: Rollback to checkpoint /org/freedesktop/NetworkManager/Checkpoint/1
      finished\nTraceback (most recent call last):\n  File \"/usr/bin/nmstatectl\",
      line 11, in <module>\n    load_entry_point('nmstate==0.3.7', 'console_scripts',
      'nmstatectl')()\n  File \"/usr/lib/python3.6/site-packages/nmstatectl/nmstatectl.py\",
      line 69, in main\n    return args.func(args)\n  File \"/usr/lib/python3.6/site-packages/nmstatectl/nmstatectl.py\",
      line 279, in apply\n    args.save_to_disk,\n  File \"/usr/lib/python3.6/site-packages/nmstatectl/nmstatectl.py\",
      line 308, in apply_state\n    save_to_disk=save_to_disk,\n  File \"/usr/lib/python3.6/site-packages/libnmstate/netapplier.py\",
      line 73, in apply\n    _apply_ifaces_state(plugins, net_state, verify_change,
      save_to_disk)\n  File \"/usr/lib/python3.6/site-packages/libnmstate/netapplier.py\",
      line 123, in _apply_ifaces_state\n    _verify_change(plugins, net_state)\n  File
      \"/usr/lib/python3.6/site-packages/libnmstate/netapplier.py\", line 136, in
      _verify_change\n    net_state.verify(current_state)\n  File \"/usr/lib/python3.6/site-packages/libnmstate/net_state.py\",
      line 63, in verify\n    self._ifaces.verify(current_state.get(Interface.KEY))\n
      \ File \"/usr/lib/python3.6/site-packages/libnmstate/ifaces/ifaces.py\", line
      446, in verify\n    cur_iface.state_for_verify(),\nlibnmstate.error.NmstateVerificationError:
      \ndesired\n=======\n---\nname: eth1\ntype: ethernet\nstate: up\nstasdte: up\n\ncurrent\n=======\n---\nname:
      eth1\ntype: ethernet\nstate: up\nethernet:\n  auto-negotiation: false\nipv4:\n
      \ enabled: false\nipv6:\n  enabled: false\nlldp:\n  enabled: false\nmac-address:
      52:55:00:D1:56:00\nmtu: 1500\n\ndifference\n==========\n--- desired\n+++ current\n@@
      -2,4 +2,13 @@\n name: eth1\n type: ethernet\n state: up\n-stasdte: up\n+ethernet:\n+
      \ auto-negotiation: false\n+ipv4:\n+  enabled: false\n+ipv6:\n+  enabled: false\n+lldp:\n+
      \ enabled: false\n+mac-address: 52:55:00:D1:56:00\n+mtu: 1500\n\n\n
```

**Current (python traceback, DEBUG, WARNING messages removed, fixed formatting)**
```
message: |
      error reconciling NodeNetworkConfigurationPolicy at desired state apply: ,
       failed to execute nmstatectl set --no-commit --timeout 480: 'exit status 1'
      libnmstate.error.NmstateVerificationError:
      desired
      =======
      ---
      name: eth1
      type: ethernet
      state: up
      stasdte: up
      current
      =======
      ---
      name: eth1
      type: ethernet
      state: up
      ethernet:
      auto-negotiation: false
      ipv4:
      enabled: false
      ipv6:
      enabled: false
      lldp:
      enabled: false
      mac-address: 52:55:00:D1:56:00
      mtu: 1500
      difference
      ==========
      --- desired
      +++ current
      @@ -2,4 +2,13 @@
      name: eth1
      type: ethernet
      state: up
      -stasdte: up
      +ethernet:
      +  auto-negotiation: false
      +ipv4:
      +  enabled: false
      +ipv6:
      +  enabled: false
      +lldp:
      +  enabled: false
      +mac-address: 52:55:00:D1:56:00
      +mtu: 1500
```

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by phoracek — CHANGES_REQUESTED** [review]
Since this is heavy on regexes, please add a set of unit tests. It would also make the PR self-documenting in a way, so it is clear to everybody what exactly has changed

**Comment by qinqon** [comment]
@tompsota can you add some output samples at the PR description ?

**Comment by phoracek** [comment]
And looking at this again, we must add e2e tests. Otherwise we risk that once nmstatectl changes it's output format, we won't be able to catch that.

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 1]
```diff
@@ -298,6 +299,76 @@ func (r *NodeNetworkConfigurationPolicyReconciler) Reconcile(ctx context.Context
 	return ctrl.Result{}, nil
 }
 
+// The function strips unnecessary parts of the error messages so it is easier to read
```
[the remark this answers is not in the document]
Comment (qinqon) [reply controllers/nodenetworkconfigurationpolicy_controller.go:302]: I am +1 on moving to a separate file, but `pkg/helper` since it's a bad package name that we will try to refactor in the futre and split it into multiple packages/files.

**Comment by qinqon** [comment]
/hold

I am not sure we want to do this, the striped part are needed for nmstate team to debug the issue.

**Comment by phoracek** [comment]
NNCE output is not for bug reporting and debugging, it is meant for the user to signalize issues in configuraiton. Errors we care about there are those that a user can resolve, for example non-existent NIC reference or a typo in bond mode.

I agree that we must keep it simple to find the traceback in case we have a bug (not a misconfiguration), but I don't think it should be in this user-facing object.

Could we just make sure that the traceback ends up in logs? To make it simple to find it, we can add a "key" for correlation - if NNCE generation number is X, I should be able to look for "nnceGeneration": "X" in logs and find the traceback there.

@qinqon does that make sense?

**Comment by qinqon** [comment]
> NNCE output is not for bug reporting and debugging, it is meant for the user to signalize issues in configuraiton. Errors we care about there are those that a user can resolve, for example non-existent NIC reference or a typo in bond mode.
> 
> I agree that we must keep it simple to find the traceback in case we have a bug (not a misconfiguration), but I don't think it should be in this user-facing object.
> 
> Could we just make sure that the traceback ends up in logs? To make it simple to find it, we can add a "key" for correlation - if NNCE generation number is X, I should be able to look for "nnceGeneration": "X" in logs and find the traceback there.
> 
> Overall, the goal of this is to enable users to be able to resolve their issues themselves without opening BZs for us to decode the cryptic output.
> 
> @qinqon does that make sense?
> 

Works for me.

> We'll see if the regex needs to be changed - e2e would fail. If it is not stable enough, we must bring this issue up again with the nmstate team and ask them to provide a real API for error reporting.
>

If we are covered by e2e tests we will see if it breaks after nmstate bumping, so I suppose we are good.

File: test/e2e/handler/broken_configuration_test.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+	Context("with invalid field", func() {
+		var (
+			messagesToKeep = [...]string{
+				"libnmstate.error.NmstateVerificationError:",
+				"desired",
+			}
+		)
+		BeforeEach(func() {
+			By("Creating the policy")
+			err := setDesiredStateWithPolicyAndNodeSelector(DefaultNetworkBroken, createInterfaceWithInvalidField(), map[string]string{"kubernetes.io/hostname": "node01"})
```
Comment (rhrazdil) [inline test/e2e/handler/broken_configuration_test.go:60]: You could define the nodeSelector variable within the Describe block, as it's the same for each test.
Also, please pick the node as `nodes[0]`, which is the first worker node in the cluster.
Running these policies on master may cause unintended trouble.

File: test/e2e/handler/broken_configuration_test.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+			By("Waiting until the node becomes ready again")
+			waitForNodesReady()
+
+			By("Waiting for policy to be ready")
+			waitForDegradedPolicy(DefaultNetworkBroken)
+
+		})
+
+		AfterEach(func() {
```
Comment (rhrazdil) [inline test/e2e/handler/broken_configuration_test.go:152]: If I'm not mistaken, the `AfterEach` is always the same, you may just define it one level higher (in the Describe block), so that it's applied for all `Context`s

File: test/e2e/handler/broken_configuration_test.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+		})
+	})
+
+	Context("with ping fail", func() {
+		var (
+			messagesToKeep = [...]string{
+				"failed to retrieve default gw at runProbes",
+			}
+		)
+		BeforeEach(func() {
```
Comment (rhrazdil) [inline test/e2e/handler/broken_configuration_test.go:137]: You may want to consider extracting the code in `BeforeEach` to a helper function taking the policy state as a parameter, since all the rest of what `BeforeEach` does is the same

File: test/e2e/handler/broken_configuration_test.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    state: up
+    ipv4:
+      address:
+      - ip: "192.168.45.33"
+        prefix-length: 24
+      dhcp: false
+      enabled: true`)
+}
+
+var _ = FDescribe("NodeNetworkState", func() {
```
Comment (rhrazdil) [inline test/e2e/handler/broken_configuration_test.go:36]: Don't forget to remove this `F` when pushing

**Review by rhrazdil — CHANGES_REQUESTED** [review]

File: test/e2e/handler/broken_configuration_test.go (hunk 3 above)
Comment (tompsota) [reply test/e2e/handler/broken_configuration_test.go:152]: I removed this part completely as it caused problems

File: test/e2e/handler/broken_configuration_test.go (hunk 2 above)
Comment (tompsota) [reply test/e2e/handler/broken_configuration_test.go:60]: Exactly this doesn't work, I'm not able to use `nodes[0]` in `Describe` block. 
However, I extracted this code into a helper function, so so the duplicity was removed anyway.

File: controllers/nodenetworkconfigurationpolicy_utils.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+)
+
+// The function strips disarranged parts of the error messages, so it is easier to read for the user
+func formatErrorString(err error) string {
+	var sb strings.Builder
+	errorMessage := err.Error()
+	errorLines := strings.Split(errorMessage, "\n")
+	index := 0
+
+	re := regexp.MustCompile(`\s*(.*--no-commit --timeout \d+: 'exit status \d+') '{2}\s'`)
```
Comment (qinqon) [inline controllers/nodenetworkconfigurationpolicy_utils.go:32]: Stuff like this is very fragile, imaging that we change the arguments order and we do --timeout and then --no-commit

File: controllers/nodenetworkconfigurationpolicy_utils.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	if index >= len(*lines) {
+		return index
+	}
+
+	regexMatchFileTrace := `\A\s*File "\S*",\sline\s\d*,\sin`
+	regexTraceback := `\A\s*Traceback\s\(most\srecent\scall\slast\):`
+	regexWithDateTime := `\A\s*\d{4}-\d{2}-\d{2}\s\d{2}:\d{2}:\d{2}`
+	regexCurrentState := `.*->\scurrentState:\s---`
+	regexUnhandledMessage := `\AUnhandled\s.*\sfor\s`
+	regexKeywords := `DEBUG|WARNING|UserWarning:|warnings\.warn\(`
```
Comment (qinqon) [inline controllers/nodenetworkconfigurationpolicy_utils.go:63]: To be honest I think all this ha to be done at nmstate itself with an output option that shows a simplified error at output and a debug error at a file for example or simplied error at stdout and extensive at stderr.

Doing it at nmstate and adding test will ensure that no new version of nmstate is going to break us.

**Review by qinqon — CHANGES_REQUESTED** [review]
I think this has to be done at nmstate and not by using regex.

**Comment by qinqon** [comment]
@tompsota after talking a little with @phoracek I am ok with this PR as long as it includes also a base64 field with the unparsed output so users don't have to dig into handler logs to put proper debug info at knmstate or nmstate bz.

**Comment by tompsota** [comment]
> @tompsota after talking a little with @phoracek I am ok with this PR as long as it includes also a base64 field with the unparsed output so users don't have to dig into handler logs to put proper debug info at knmstate or nmstate bz.

Allright, I'll add the field then.

**Comment by rhrazdil** [comment]
@tompsota the issue with the failing test that creates policy setting wrong IP address on a node should be fixed in this PR https://github.com/nmstate/kubernetes-nmstate/pull/786

File: test/e2e/handler/error_messages_formatting_test.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+func createInterfaceWithInvalidField() nmstate.State {
+	return nmstate.NewState(`interfaces:
+  - name: eth1
+    type: ethernet
+    stasdte: up`)
+}
+
+func createInterfaceWithPingFail() nmstate.State {
```
Comment (rhrazdil) [inline test/e2e/handler/error_messages_formatting_test.go:24]: Maybe better to say `createInterfaceWIthIncorrectIP`

File: test/e2e/handler/error_messages_formatting_test.go [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+func createInterfaceWithInvalidName() nmstate.State {
+	return nmstate.NewState(`interfaces:
+  - name: eth666
+    type: ethernet
+    state: up`)
+}
+
+func createInterfaceWithInvalidField() nmstate.State {
+	return nmstate.NewState(`interfaces:
+  - name: eth1
```
Comment (rhrazdil) [inline test/e2e/handler/error_messages_formatting_test.go:19]: Better use eth0, so that it doesn't depend on nodes having secondary ifaces

File: test/e2e/handler/error_messages_formatting_test.go [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+		})
+
+		It("should keep desired parts of the message", func() {
+			for _, failMessage := range messagesToKeep {
+				Expect(enactmentConditionsStatus(nodes[0], defaultPolicy).Find(nmstate.NodeNetworkConfigurationEnactmentConditionFailing).Message).To(ContainSubstring(failMessage))
+			}
+		})
+	})
+
+	Context("with invalid interface name", func() {
```
Comment (rhrazdil) [inline test/e2e/handler/error_messages_formatting_test.go:94]: ditto

File: test/e2e/handler/error_messages_formatting_test.go [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+		AfterEach(func() {
+			resetDesiredStateForNodes()
+			By("Remove the policy")
+			deletePolicy("test-policy")
+		})
+
+		It("should discard disarranged parts of the message", func() {
+
+			for _, failMessage := range messagesToRemove {
```
Comment (rhrazdil) [inline test/e2e/handler/error_messages_formatting_test.go:139]: maybe unwantedMessage/desiredMessage would be a better name for these variables?

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 12]
```diff
@@ -184,16 +184,22 @@ func (r *NodeNetworkConfigurationPolicyReconciler) Reconcile(ctx context.Context
...
+	nmstateOutput, err := nmstate.ApplyDesiredState(r.APIClient, policyInstance.Spec.DesiredState)
 	if err != nil {
 		errmsg := fmt.Errorf("error reconciling NodeNetworkConfigurationPolicy at desired state apply: %s, %v", nmstateOutput, err)
 		encodedMessage, _ := encodeMessage(err.Error())
-		strippedErrmsg := fmt.Errorf("error reconciling NodeNetworkConfigurationPolicy at desired state apply: %s,\n %s \n %s", nmstateOutput, formatErrorString(err), encodedMessage)
+		strippedErrmsg := fmt.Errorf("error reconciling NodeNetworkConfigurationPolicy at desired state apply: %s,\n %s", nmstateOutput, formatErrorString(err))
 
 		enactmentConditions.NotifyFailedToConfigure(strippedErrmsg)
+
+		err := r.updateEncodedMessage(policyInstance, encodedMessage)
```
Comment (rhrazdil) [inline controllers/nodenetworkconfigurationpolicy_controller.go:198]: Could we also do it here? Assuming `SetFailedToConfigure` handles compression and b64 encoding of course
https://github.com/nmstate/kubernetes-nmstate/blob/a51f39f9152e15ff89fb01ac5e092fcc5bf3a724/pkg/enactmentstatus/conditions/conditions.go#L103

**Review by rhrazdil — CHANGES_REQUESTED** [review]
Thanks!
It looks like a lot of comments, but most are just minor things

File: controllers/nodenetworkconfigurationpolicy_utils.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
+/*
```
Comment (rhrazdil) [inline controllers/nodenetworkconfigurationpolicy_utils.go]: Could you please move this file to `pkg/enactment/`? That should be a better home for it.
Also, a better name would be nice, `message.go` maybe?

File: api/shared/nodenetworkconfigurationpolicy_utils_test.go [hunk 14]
```diff
@@ -0,0 +1,75 @@
+package shared
```
Comment (rhrazdil) [inline api/shared/nodenetworkconfigurationpolicy_utils_test.go:1]: similarly, `pkg/enactment/message_test.go`

File: api/shared/nodenetworkconfigurationpolicy_utils_test.go [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
```
Comment (rhrazdil) [inline api/shared/nodenetworkconfigurationpolicy_utils_test.go:2]: I'm wondering now if these tests are even executed, I fear they might not.
I believe you're missing a suite file, take a look at for example `pkg/names/names_suite_test.go`, this file is needed for each package to have the tests executed

**Review by rhrazdil — CHANGES_REQUESTED** [review]

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 16]
```diff
@@ -190,8 +190,10 @@ func (r *NodeNetworkConfigurationPolicyReconciler) Reconcile(ctx context.Context
 	nmstateOutput, err := nmstate.ApplyDesiredState(r.APIClient, instance.Spec.DesiredState)
 	if err != nil {
 		errmsg := fmt.Errorf("error reconciling NodeNetworkConfigurationPolicy at desired state apply: %s, %v", nmstateOutput, err)
+		encodedMessage, _ := encodeMessage(err.Error())
```
Comment (rhrazdil) [inline controllers/nodenetworkconfigurationpolicy_controller.go]: Question: shouldn't this be:
```
encodedMessage, _ := encodeMessage(errmsg)
```

File: pkg/enactmentstatus/message.go [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+	re := regexp.MustCompile(`\s*(failed to execute.*'exit status \d+') '{2}\s'`)
+	matches := re.FindStringSubmatch(errorLines[index])
+	if len(matches) == 2 {
+		sb.WriteString(strings.TrimRight(matches[1], " ") + "\n")
+		index++
+	}
+	index = skipLines(&errorLines, index)
+
+	for index < len(errorLines) {
+		lineSplitByColon := strings.Split(strings.TrimRight(errorLines[index], " "), ": ")
```
Comment (rhrazdil) [inline pkg/enactmentstatus/message.go:48]: Could the code in this for loop abstracted by a function?

File: pkg/enactmentstatus/message_test.go [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+			Expect(FormatErrorString(pingInvalidInput)).To(Equal(pingInvalidOutput))
+		})
+		It("Should keep message", func() {
+			Expect(FormatErrorString(pingValidInput)).To(Equal(pingValidOutput))
+		})
+	})
+
+	Describe("Error messages compression and decompression", func() {
+		var (
+			messages = [...]string{
```
Comment (rhrazdil) [inline pkg/enactmentstatus/message_test.go:58]: Why not just use slice?
`[]string{"a", "b", ...}`

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 19]
```diff
@@ -121,8 +121,8 @@ func (r *NodeNetworkConfigurationPolicyReconciler) Reconcile(ctx context.Context
 	log := r.Log.WithValues("nodenetworkconfigurationpolicy", request.NamespacedName)
 
 	// Fetch the NodeNetworkConfigurationPolicy instance
-	instance := &nmstatev1beta1.NodeNetworkConfigurationPolicy{}
-	err := r.Client.Get(context.TODO(), request.NamespacedName, instance)
+	policyInstance := &nmstatev1beta1.NodeNetworkConfigurationPolicy{}
```
Comment (rhrazdil) [inline controllers/nodenetworkconfigurationpolicy_controller.go:124]: the variable name change is not explained in the commit message, I'm not sure if it's needed.
Please either put this change in a separate commit of keep the original

Comment (tompsota) [reply controllers/nodenetworkconfigurationpolicy_controller.go:124]: It's not necessary, I'll remove this change

File: controllers/nodenetworkconfigurationpolicy_controller.go (hunk 16 above)
Comment (tompsota) [reply controllers/nodenetworkconfigurationpolicy_controller.go]: This code was slightly modified and moved in the last commit into `SetFailedToConfigure` function in `pkg/enactmentstatus/conditions/conditions.go` and it works as expected (it is encoded also with the introductory string)

File: pkg/enactmentstatus/message.go [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+	}
+	index = skipLines(&errorLines, index)
+
+	selectAndFormatLines(index, &errorLines, &sb)
+
+	return sb.String()
+}
+
+//Loops over all lines in error message and selects lines that should be kept
+func selectAndFormatLines(index int, errorLines *[]string, sb *strings.Builder) {
```
Comment (rhrazdil) [inline pkg/enactmentstatus/message.go:53]: Could we name this just `formatLines`?
The select part seems redundant

File: pkg/enactmentstatus/message.go [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+	for index < len(*errorLines) {
+		formatCurrentLine(&(*errorLines)[index], sb)
+		index++
+		index = skipLines(errorLines, index)
+	}
+}
+
+//Simplifies a line that should be kept in the error message
+func formatCurrentLine(errorLine *string, sb *strings.Builder) {
```
Comment (rhrazdil) [inline pkg/enactmentstatus/message.go:62]: And this would read better if only called `formatLine`

File: pkg/enactmentstatus/message.go [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+	lineSplitByColon := strings.Split(strings.TrimRight(*errorLine, " "), ": ")
+	indent := ""
+	for _, lineSection := range lineSplitByColon {
+		sb.WriteString(indent + lineSection + "\n")
+		indent = indent + "  "
+	}
+}
+
+func skipLines(lines *[]string, index int) int {
```
Comment (rhrazdil) [inline pkg/enactmentstatus/message.go:71]: nit: IIUC when passing a slice, go passes the slice header, which points to the underlying array.
So it should not be neccessary to pass a pointer to the slice here

**Review by rhrazdil — CHANGES_REQUESTED** [review]
Thanks, only coments regarging the names of the introduced functions

**Comment by rhrazdil** [comment]
Thanks!
/approve

**Review by phoracek — COMMENTED** [review]
/lgtm
/hold cancel
